### DISCUSSION TITLE
Product Function Consultation

### DISCUSSION THREAD

**Discussion by ac169** [savoirfairelinux/opendht#729] [Q&A]
1. Can nodes using different bootstrap nodes communicate with each other
2. Does communication between nodes have to be forwarded by bootstrap nodes or can it be self penetrated for point-to-point communication

**Comment by aberaud** [comment]
1. yes, any node in the network can be used as a bootstrap node. Suggested bootstrap nodes are just normal OpenDHT nodes with a known and stable public address.
2. no, nodes in OpenDHT communicate directly in peer to peer. DHT nodes (including bootstrap nodes) don't forward messages, they only exchange lists of DHT peers (and values). Bootstrap nodes are only used to know about other nodes on startup, when we don't yet know about any other nodes in the network.
- It's possible for a node to persist a cache of known nodes to be used, instead or additionally to bootstrap nodes, for following runs.
- It's also possible to use automatic multicast peer discovery to start a DHT on a local network even without any bootstrap node.

**Reply by ac169** [reply]
**thank you!**

I don't think the first question was described clearly. Let me give an example of my question:

The names of the two bootstrap nodes are: bootA bootB

There are currently ten client nodes, clinet1, clinet2, clinet3 ... clinet10, Can these 10 client segments communicate with each other, with clipet1-5 starting using bootA and the remaining clipet6-10 starting using bootB?

**Reply by aberaud** [reply]
Yes, as long as bootA and bootB are part of the same DHT network.

**Reply by aberaud** [reply]
Bootstrap nodes are only used to initially know about other peers in the dht network, when the node routing table is empty. This is why any node can be used as a bootstrap to connect to a dht network. Every node in this network are able to exchange values.

**Reply by ac169** [reply]
> Bootstrap nodes are only used to initially know about other peers in the dht network, when the node routing table is empty. This is why any node can be used as a bootstrap to connect to a dht network. Every node in this network are able to exchange values.

```
dhtchat -b bootstrap.jami.net
```

Using the examples in the source code for testing, it was found that in both IPv4 and IPv6 network environments, it was not the point-to-point direct communication you mentioned, but rather forwarded by third-party servers such as (2001:41d0: *: *:: 149.202. *. * 31.207. *. *)! No part related to NAT penetration was found in the code

**Reply by aberaud** [reply]
OpenDHT is a distributed network providing a global, distributed key-value storage. It's a lightweight, UDP-based network and doesn't include NAT-traversal. It's designed and work best for small, volatile data.

Node don't send values to each other directly, they use other nodes on the DHT network to store data. Every node in the network can store data for other nodes by default.

So node A that performs `put(H, value)` will store data on other random nodes on the network (those whose ID happens to be close to H).
Node B performing `get(H)` or `listen(H)` will query these same nodes for values. Node A and B can randomly be picked by other nodes to store data too.

If you want to establish direct, peer-to-peer/point to point connections, you can use DHTNet ( https://github.com/savoirfairelinux/dhtnet ) which makes use of OpenDHT to exchange [ICE](https://en.wikipedia.org/wiki/Interactive_Connectivity_Establishment) candidates ([RFC 8445](https://datatracker.ietf.org/doc/html/rfc8445), [RFC 6544](https://datatracker.ietf.org/doc/html/rfc6544) etc.) and perform TCP NAT traversal, and then establishes an authenticated peer-to-peer TLS connection.

**Reply by ac169** [reply]
> OpenDHT is a distributed network providing a global, distributed key-value storage. It's a lightweight, UDP-based network and doesn't include NAT-traversal. It's designed and work best for small, volatile data.
> 
> Node don't send values to each other directly, they use other nodes on the DHT network to store data. Every node in the network can store data for other nodes by default.
> 
> So node A that performs `put(H, value)` will store data on other random nodes on the network (those whose ID happens to be close to H). Node B performing `get(H)` or `listen(H)` will query these same nodes for values. Node A and B can randomly be picked by other nodes to store data too.
> 
> If you want to establish direct, peer-to-peer/point to point connections, you can use DHTNet ( https://github.com/savoirfairelinux/dhtnet ) which makes use of OpenDHT to exchange [ICE](https://en.wikipedia.org/wiki/Interactive_Connectivity_Establishment) candidates ([RFC 8445](https://datatracker.ietf.org/doc/html/rfc8445), [RFC 6544](https://datatracker.ietf.org/doc/html/rfc6544) etc.) and perform TCP NAT traversal, and then establishes an authenticated peer-to-peer TLS connection.

Perhaps there is no big problem if the existing source code is only used on Linux, but for developers who want to expand functions based on these two projects, they all have one thing in common: cross-platform and development ease-of-use are very imperfect. Developers It takes a lot of time to read the code and do dependency adaptation.

**Reply by aberaud** [reply]
Thanks for the feedback
Please open a dedicated ticket if you have issue building for any supported platform
Thank you
